src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,413 @@
+            if (bbh_ == bad_batch_handling::error) {
+                throw std::runtime_error{msg};
+            }
+            logger::warn(msg);
+        }
+        return false;
+    }
+    return true;
+}
+
+bool
+image_reader::crop(cv::Mat &src, cv::Mat &dst, instance const &ins) const
+{
+    if (src.rows < img_dims_[1] || src.cols < img_dims_[2]) {
+        if (bbh_ != bad_batch_handling::skip) {
+            auto msg = fmt::format(
+                "Input image dimensions [rows: {0:n}, cols: {1:n}] are "
+                "smaller than the output image image_dimensions [rows: {2:n}, "
+                "cols: {3:n}], for the image: {4:n} in data_store: {5:s}",
+                src.rows,
+                src.cols,
+                img_dims_[1],
+                img_dims_[2],
+                ins.index(),
+                ins.get_data_store().id());
+            if (bbh_ == bad_batch_handling::error) {
+                throw std::invalid_argument{msg};
+            }
+            logger::warn(msg);
+        }
+        return false;
+    }
+
+    // center crop
+    int y = (src.rows - img_dims_[1]) / 2;
+    int x = (src.cols - img_dims_[2]) / 2;
+
+    try {
+        cv::Rect roi(x, y, img_dims_[2], img_dims_[1]);
+        src(roi).copyTo(dst);
+    }
+    catch (cv::Exception const &e) {
+        if (bbh_ != bad_batch_handling::skip) {
+            auto msg = fmt::format(
+                "Image crop operation failed for the image: {0:n} in "
+                "data_store: {1:s}, with the following exception: {2:s}",
+                ins.index(),
+                ins.get_data_store().id(),
+                e.what());
+            if (bbh_ == bad_batch_handling::error) {
+                throw std::runtime_error{msg};
+            }
+            logger::warn(msg);
+        }
+        return false;
+    }
+    return true;
+}
+
+}  // namespace v1
+}  // namespace mlio
+
+#else
+
+// clang-format off
+
+#include "mlio/not_supported_error.h"
+
+namespace cv {
+class Mat {};
+}  // namespace cv
+
+namespace mlio {
+inline namespace v1 {
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmissing-noreturn"
+
+image_reader::image_reader(data_reader_params rdr_prm,
+                           image_reader_params image_prm)
+    : parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}
+{
+    bbh_ = {};
+    img_dims_ = {};
+    throw not_supported_error{"ML-IO has not been built with image reader "
+                              "support."};
+}
+
+image_reader::~image_reader() = default;
+
+intrusive_ptr<record_reader>
+image_reader::make_record_reader(data_store const &)
+{
+   return nullptr;
+}
+
+intrusive_ptr<schema const>
+image_reader::infer_schema(std::optional<instance> const &)
+{
+    return nullptr;
+}
+
+intrusive_ptr<example>
+image_reader::decode(instance_batch const &) const
+{
+    return nullptr;
+}
+
+intrusive_ptr<dense_tensor>
+image_reader::make_tensor(std::vector<instance> const &,
```

Comment:
Minor. Both `make_tensor` and `decode_image` can fit into a single line as they don't have argument names.
```
image_reader::make_tensor(std::vector<instance> const &, std::size_t) const
{
...
}
```